### ISSUE TITLE
Feature request: better explanatory text about location permissions

### ISSUE THREAD

**Issue by saalkom** [AU-COVIDSafe/mobile-android#15]
It would be a great idea if an explanation for why location permissions are needed at all is clearly given somewhere at the top of the main screen near where the "COVIDSafe does not send pairing requests" message is, e.g.:

> COVIDSafe doesn't send your location to the government, but all Android apps using the same technology COVIDSafe uses must still request location permissions/your location(?). Tap here to read why.

**_If_** I understand the reason correctly, the explanation could be along the lines of:

> COVIDSafe is based on technology which uses Bluetooth Low Energy (BLE) to detect other devices running COVIDSafe. Some companies/organizations(?) also use BLE to deliver targeted advertising/messaging(?) to users' devices as they move around through a building — for example, department stores may send tailored advertisements to customers who have that store's app installed as they move between parts of a store. Using BLE that way may raise location privacy issues, and so the developers in charge of Android force all apps using Bluetooth/BLE(?) to also explicitly request permission to access the user's location. COVIDSafe does not send advertisements and does not provide your location data either to the government or to any health authority. For technical details see https://developer.android.com/guide/topics/connectivity/bluetooth

If Google location services (or other similar geolocation provider) must be or may inadvertently be enabled to use all COVIDSafe's features, then I'm not sure you could in good faith say "...and does not provide your location data [...] _to any third party_" in the disclaimer at the end. That's a legal question I guess.

Perhaps an example in the explanation which doesn't mention advertising at all (and thus wouldn't insinuate to lazy readers that COVIDSafe sends ads) would be better, but the point is that a government app requesting location permissions instinctively gives users some pause — especially when there were previously privacy concerns about said app that dominated news headlines for days or more, and even more so when the app in question superficially seems to be doing _exactly_ what Australians were afraid it would do and were reassured it would _not_ do — and is probably in no small part responsible for inadequate app uptake, app ineffectiveness, and app uninstallations. While I try to keep myself educated and do have implicit faith in software auditors and independent paranoids fluent in Java and Kotlin to review the source code and keep the government to its privacy commitments, the general public is neither that technically-minded nor familiar with the benefits and implications of the open-source model. If the COVIDSafe app is going to continue to be a centrepiece of Australia's coronavirus/COVID-19 response and recovery, then there simply needs to be a conspicuous and candid (but not overly-technical) explanation for why location permissions are needed and why location needs to be enabled and stay enabled. The current "Location permission for Android" explainer buried in the help menu amounts to a mere half-measure in light of public privacy concerns about the COVIDSafe app in combination with the sheer emphasis that federal, state and territory governments are placing on it.

While you're at it, put an explainer on the Play Store page too. At the time of writing, over there your app was sitting at a mere 2.9 stars out of five.

**Comment by covidsafe-support** [comment]
Hi @saalkom - thanks for the suggestion. We've added this to our backlog to look in to further.

**Comment by saalkom** [comment]
Thank you. I just want there to be public trust in the app so that the contact-tracing effort isn't hindered.

**Comment by wallarug** [comment]
Hey @saalkom ,

Do you know why the location permission is now mandated for the Android App as of version 1.0.48?  

I just noticed the app updated and now wants Location turned on all the time to be 'happy'.  I have read through the other issues and note that the app still scans when location is off, is this correct?

**Comment by jimmo** [comment]
hi @wallarug -- see this issue that I rasised https://github.com/AU-COVIDSafe/mobile-android/issues/14 which explains what's happening here.

In summary, the app always required location permission. You would had previously given the app "fine" location permission (likely implicitly during installation). However, as of v1.0.39 it now requires "coarse" location permission instead (but doesn't use your existing "fine" permission that it already has).

The app needs location permission in order to use some Bluetooth functionality (as technically you can infer location by detecting a BLE location beacon). COVIDSafe definitely doesn't use any other location (e.g. GPS) functionality though.

Many users are being updated directly to v1.0.48 from much older versions, so it's possible you never had v1.0.39. This is because, in many cases, the app was previously unable to auto-update (even if you have auto-updates enabled), however this appears to be fixed as of v1.0.48.

**Comment by wallarug** [comment]
Thanks @jimmo.  I understand what you are saying about all the permission stuff now.

So does the Location toggle have to be switched on for the app to function now as well or just granting the permission and ignoring the warnings suffice?

**Comment by jimmo** [comment]
@wallarug The global "Location" enabled setting sort of acts like a global kill switch for the location permission in all apps.

This app does two things simultaneously - scan for and connect to other phones, and broadcast presence to and receive connections from other phones. Whenever a connection is made (in either direction), both phones exchange eachother's tracing IDs and store them.

However, the UI here is quite confusing - it's easy to be jn a state where the app says it isn't active but it's still detectable by other phones. Also the notification doesn't always say the same thing as the main screen in the app. See https://github.com/AU-COVIDSafe/mobile-android/issues/13#issuecomment-663328754 for the specifics.

To answer your question though, in order for the app to be fully functional (i.e. able to detect other phones running COVIDSafe) then you must have location enabled and the location permission granted.

If location is disabled, then the app will still be detectable by other phones (and they connect to you). As all exchanges are bidirectional, this means that you'll successfully encounter log the other phones (and they'll log you), but of course this requires that everyone else has location enabled so that they can detect and connect to you.

**Comment by kurahaupo** [comment]
In the subject of needing to improve explanations, I initially installed the app, registered, and then disabled its access to the internet, on the assumption that it was only (legitimately) needed for registration and infection notification.

I subsequently heard something about it needing to fetch new keys periodically, but no explanation of why that was necessary, both in terms of how they are used, and in terms of why it was decided to adopt a design that works in this way, including any tradeoffs that were considered, or could be considered post-hoc.

**Comment by crispyoz** [comment]
This is exactly why myself and most of my colleagues removed the app from our devices. When the minister states that the application is "totally secure", you know he doesn't know what he is talking about.   While the app does not record location data, the fact that location is required to be turned on in order for the app to function, makes your device location available to google and others to track your movement. That is a fundamental security flaw.

**Comment by kurahaupo** [comment]
@crispyoz agreed that the minister probably has no idea what they're talking about.

However what matters isn't the minister's opinions but the actual technical details. There _are_ valid reasons for needing the permissions, it's the insufficiency of technical explanation that's the issue.

I'd like to know more about:
1. (how) does rotating the device keys improve my security by making it harder for other actors to track my device using a modified version of the app?
2. does turning off GPS actually appear as "permission revoked" to apps that want to exchange bluetooth beacons, both (a) when checking permissions, and (b) when attempting bluetooth operations?

**Comment by kurahaupo** [comment]
@saalkom closed, but with what resolution?

**Comment by saalkom** [comment]
@kurahaupo You're right; the new explainer ("Android apps need location permissions turned on to access Bluetooth. COVIDSafe does not store or use location data.") amounts to a mere half-measure and I shouldn't have closed this issue.

**Comment by saalkom** [comment]
There is also still no explanation at all for location permissions on the Play Store page.

**Comment by vteague** [comment]
The rich irony here is that this is one of very few ways in which COVIDSafe's privacy properties are clearly superior to those of the G/Apple API for Android users.  As @jimmo explained above, COVIDSafe can still function successfully even when the global location setting is turned off - it can advertise, and exchange in both directions, as long as it doesn't meet another paranoid who (like me) also has their global location turned off.

Unfortunately, the UI says that it isn't working at all, when in fact it is mostly working.  And the online explanations are - as many have remarked above - worse than useless.

By contrast, the current implementation of the G/Apple API on Android requires the global location setting to be turned on, so it doesn't work at all unless you agree to share your location with Google constantly.  Google have promised to fix this in Android 11, but that won't help most of us for quite some time.

https://support.google.com/android/answer/9930236

**Comment by kurahaupo** [comment]
> [...] so it doesn't work at all unless you agree to share your location with Google constantly.

That would be a gross misrepresentation of how Android permissions work.

Your phone's operating system always knows its location, unless you turn off the sensors. The permissions have nothing to do with that. What the permissions do determine is whether or not each app can obtain that information from the OS.

Some apps provided by Google do indeed store your location data on Google's services, but there's no way that granting or withholding a permission to AU-COVIDSafe can change whether anyone besides covid.gov.au can see your location.

And you can look at the source code to satisfy yourself that it's not going to sense your location without your explicit confirmation.

**Comment by vteague** [comment]
My remark was quite clearly about the global location setting on Android, not about the specific location permission for the COVIDSafe app. 

If I you think I have mischaracterised the effect of switching on Android's global location setting, given that I otherwise always have it turned off, and do not think that turning it on informs Google of my location, while leaving it off generally gives Google only coarser information about my location, then you are welcome to explain politely why you think I am wrong.  But please read what I wrote before you write "gross misrepresentation" or similar again, thanks.

And btw I have inspected the COVIDSafe Android source code already, with some diligence.

**Comment by kurahaupo** [comment]
@vteague OK, that wasn't fair of me, sorry. I'm a bit trigger-happy about people making sensationalist claims about Android, and I didn't read what you'd said carefully enough.

That said, Google probably records your location information, but that's not _just_ because the location sensor is turned on.

It's my understanding that Google-supplied apps don't get a magic bypass on the permissions; rather they just ask for (and are pre-granted) a large number of permissions, and _using those permissions_, Google gets access to your location. Revoke those, and you're clear. Obviously apps like Maps won't be as useful without location permission, so "always off" won't suit everyone. Recent versions of Android allow permissions to be granted either "once", "always", or "only when in the foreground", which seems like a good usability compromise.

(If one doesn't trust Android not to have a magic permissions bypass for selected apps, then one shouldn't trust it to really turn off the location sensors either. But it seems like it's fairly easy to test: revoke the permissions, and if the app complains, then the permissions are almost certainly being correctly enforced. The other way to check is to root your device, and then check what's actually running as root besides your shell. I've done both of these checks and not found any evidence to contradict my previous paragraph.)

Some OEM versions of Android have a "settings" app that won't let you modify the permissions granted to pre-installed apps, but that's easily worked around by installing a third-party permissions-setting app, as the underlying OS will still let you modify the permissions.

**Comment by kurahaupo** [comment]
@saalkom I note that a recent update to `AU-CovidSAFE ` resulted a one-off pop-up explaining why location permission is required; in `app/src/main/res/values/strings.xml` is:
```lang=xml
<string name="home_set_location_why">Google requires Location permission for Android apps to access Bluetooth®. COVIDSafe does not store or use location data.</string>
```

Does that mean this issue can be closed?

**Comment by kurahaupo** [comment]
PS: why is this Github project called `AU-CovidSAFE` when
* `COVIDSafe` appears in the source code 765 times,
* `covidsafe` appears 665 times,
* `CovidSafe` appears thrice,
* `COVIDsafe` appears twice,
* `COVIdSafe` appears twice (yes, really), and
* `CovidSAFE` appears **zero** times.

**Comment by saalkom** [comment]
> @saalkom I note that a recent update to `AU-CovidSAFE ` resulted a one-off pop-up explaining why location permission is required; in `app/src/main/res/values/strings.xml` is:
> ```lang=xml
> <string name="home_set_location_why">Google requires Location permission for Android apps to access Bluetooth®. COVIDSafe does not store or use location data.</string>
> ```
> 
> Does that mean this issue can be closed?

That's not really much different from the message I had previously noted in my comment above (https://github.com/AU-COVIDSafe/mobile-android/issues/15#issuecomment-698198223), although at least there's more of a chance people will read it now that it's displayed in a popup.

However, all the discussion here has got me thinking whether the explanatory text actually just needs to specify that COVIDSafe needs location turned on to be able to record close contacts that have location turned off on their devices, rather than for the app to work at all. Is it fair to say we have a consensus about this?

> PS: why is this Github project called `AU-CovidSAFE`

Where are you reading this exactly?

**Comment by nopresnik** [comment]
> PS: why is this Github project called `AU-CovidSAFE` when
> 
> * `COVIDSafe` appears in the source code 765 times,
> * `covidsafe` appears 665 times,
> * `CovidSafe` appears thrice,
> * `COVIDsafe` appears twice,
> * `COVIdSafe` appears twice (yes, really), and
> * `CovidSAFE` appears **zero** times.
> 
> ```
> $ grep -oirh covidsafe * | sort | uniq -c
>     760 COVIDSafe
>       2 COVIDsafe
>       2 COVIdSafe
>       3 CovidSafe
>     663 covidsafe
> ```

You are confusing the project name with the Github organisation name.  It appears [COVIDSafe](https://github.com/covidsafe) has already been taken so AU-COVIDSafe is an appropriate alternative. :)

**Comment by vteague** [comment]
@kurahaupo Fair point - I hadn't adequately considered the case in which I turn on the global location setting but assiduously refuse location permission to any app that might use location services.  You're probably right about that case.  Pax.

**Comment by kurahaupo** [comment]
@saalkom 
> > PS: why is this Github project called `AU-CovidSAFE`
> 
> Where are you reading this exactly?

Sadly it would appear that I was citing my own earlier typo, so I withdraw the question.

**Comment by kurahaupo** [comment]
> > PS: why is this Github project called `AU-CovidSAFE` when
...
> You are confusing the project name with the Github organisation name.  It appears [COVIDSafe](https://github.com/covidsafe) has already been taken so AU-COVIDSafe is an appropriate alternative. :)

I'm not the slightest bit confused on that point, but I may have been confused on the correct capitalisation.

Don't you think the proliferation of upper- & lower-case versions is odd?
